designs/Strings-design.md

```diff
@@ -0,0 +1,265 @@
+    2. We are thinking about changing the IR.VarInit\<A> = { a?: A, name: string, type: Type, value: Value\<A> } to IR.VarInit\<A> = { a?: A, name: string, type: Type, value: Expr\<A> }. By this way, we can accomplish the allocation and storing jobs.
+    3. Actually, we are thinking about also change the AST.VarInit\<A> = { a?: A, name: string, type: Type, value: Literal } to AST.VarInit\<A> = { a?: A, name: string, type: Type, value: Expr\<A> }. We know that in ChocoPy, when initialzing the variables, we should not use a expression at the right hand side. Thus, when initialzing the variables which need the memory, the right hand side can only be None. However, by modifying the VarInit stucture, we can support the initializations such as l: [int] = [1, 2, 3] and c: C = C() where C is a class object. Then our behavior will be more python-like.
+
+    Our temporay solution is that we require the user of our compiler to defined a string in two steps, which is the same as defining a class:
+    1. Assign "None" to 1. (e.g. x:str = None)
+    2. Assign a value to it. (e.g. x = "abcd")
+    
+    Therefore, the test case 1 will be:
+    ```Python
+    s:str = None
+    s = "abcd"
+    print(s)
+    ```
+    Expected output:
+    ```Python
+    abcd
+    ```
+
+2. Return the length of a string.
+    ```Python
+    s:str = "abcd"
+    print(len(s))
+    ```
+    Expected output:
+    ```Python
+    4
+    ```
+
+    **Update Week 7**:
+
+    Our solution is that when storing a string, we firstly store its length on top of its memory block (The string type should be immutable, so the length will not change). By doing so, when we want to load the string, we can know where to stop when generating the wasm loop block.
+
+    Besides, regrading the builtin print and len, we would like to define two functions in wasm: print_str and len_str to handle the string case. For print_str, we first get the length of the string, and then generate a wasm loop to load the ASCIIs of the strings and convert them back to string. For len_str, simpler, we can get the length and call print_num (haven't implemented).
+
+    For now, we implement the print_str in complier.ts by adding new variables to get the length of the string and handle the loop block in wasm. In the following weeks, we should implement them in lower.ts using
+    the for loop designed by the fop loop group.
+
+    ```Python
+    s:str = None
+    s = "abcd"
+    print(len(s))
+    ```
+    Expected output:
+    ```Python
+    4
+    ```
+
+3. Concatenate two strings
+    ``` Python
+    s1:str = "ab"
+    s2:str = "cd"
+    print(s1 + s2)
+    ```
+    Expected output:
+    ```Python
+    abcd
+    ```
+
+    **Update Week 7**:
+
+    One main issue of the operations (concat, comparsion, indexing) on string is that we only have the information of the string starting address. If we want to realize these operations without touching the complier.ts, I think we need more information of string.

```

Comment:
Now we have fixed the problem. We add variants of allocate and store, now they can receive an expression as the value. The way we do this is because we don't know the length of a string in this case: **a = b[1]**. (Both a and b are strings). So, we have to use the **load** to get the value. Now, concat, comparison and index are implemented and tested.